crates/swc_plugin_runner/src/lib.rs

```diff
@@ -108,70 +108,129 @@ fn load_plugin(plugin_path: &Path, cache: &mut Option<PluginCache>) -> Result<In
+        {
+            cell.set(*byte)
+        }
 
-/// Copy guest's memory into host, construct serialized struct from raw bytes.
-/// Once copy completes, fn frees allocated memory in guest as well.
-fn read_bytes_from_guest(
-    instance: &Instance,
-    memory: &Memory,
-    returned_ptr: i32,
-    len: i32,
-) -> Result<Serialized, Error> {
-    let ptr: WasmPtr<u8, Array> = WasmPtr::new(returned_ptr as _);
+        let ptr = (allocated_ptr, serialized_len.try_into()?);
 
-    // Deref & read through plugin's wasm memory space via returned ptr
-    let derefed_ptr = ptr
-        .deref(memory, 0, len.try_into()?)
-        .ok_or(anyhow!("Failed to deref raw bytes from plugin's memory"))?;
+        self.allocated_ptr_vec.push(ptr);
+        Ok(ptr)
+    }
 
-    let transformed_raw_bytes = derefed_ptr
-        .iter()
-        .enumerate()
-        .take(len.try_into()?)
-        .map(|(_size, cell)| cell.get())
-        .collect::<Vec<u8>>();
+    /// Copy guest's memory into host, construct serialized struct from raw
+    /// bytes.
+    fn read_bytes_from_guest(&mut self, returned_ptr: i32, len: i32) -> Result<Serialized, Error> {
+        let memory = self.instance.exports.get_memory("memory")?;
+        let ptr: WasmPtr<u8, Array> = WasmPtr::new(returned_ptr as _);
+
+        // Deref & read through plugin's wasm memory space via returned ptr
+        let derefed_ptr = ptr
+            .deref(memory, 0, len.try_into()?)
+            .ok_or(anyhow!("Failed to deref raw bytes from plugin's memory"))?;
+
+        let transformed_raw_bytes = derefed_ptr
+            .iter()
+            .enumerate()
+            .take(len.try_into()?)
+            .map(|(_size, cell)| cell.get())
+            .collect::<Vec<u8>>();
+
+        Ok(Serialized::new_for_plugin(&transformed_raw_bytes[..], len))
+    }
 
-    let free_memory_in_guest = instance
-        .exports
-        .get_native_function::<(i32, i32), i32>("__free")?;
+    fn transform(
+        &mut self,
+        program: &Serialized,
+        config: &Serialized,
+    ) -> Result<Serialized, Error> {
+        let guest_program_ptr = self.write_bytes_into_guest(program)?;
+        let config_str_ptr = self.write_bytes_into_guest(config)?;
+
+        let returned_ptr = self.exported_plugin_transform.call(
+            guest_program_ptr.0,
+            guest_program_ptr.1,
+            config_str_ptr.0,
+            config_str_ptr.1,
+        )?;
 
-    free_memory_in_guest.call(returned_ptr, len)?;
+        self.allocated_ptr_vec.push(returned_ptr);
+        self.read_bytes_from_guest(returned_ptr.0, returned_ptr.1)
+    }
+}
 
-    Ok(Serialized::new_for_plugin(&transformed_raw_bytes[..], len))
+impl Drop for PluginTransformTracker {
+    fn drop(&mut self) {
+        for ptr in self.allocated_ptr_vec.iter() {
+            self.exported_plugin_free
+                .call(ptr.0, ptr.1)
+                .expect("Failed to free memory allocated in the plugin");
```

Comment:
`free` also can fail, but if that occurs it's probably happening in runtime (wasmer) side which we can't recover to proceed anyway, unwrapping via `expect`.